File: scripts/cleanup-text-messages.py
#!/usr/bin/env python3
"""
Group Text Message Cleanup Script (Python + Telethon)

This script deletes all text-only messages from a Telegram group,
except for messages sent today. Media messages are preserved.

Prerequisites:
    pip install telethon python-dotenv

Usage:
    python scripts/cleanup-text-messages.py

You'll need:
1. API ID and API Hash from https://my.telegram.org
2. Your phone number for authentication
3. Admin rights in the target group
"""

import os
import asyncio
from datetime import datetime, timedelta
from telethon import TelegramClient
from telethon.tl.types import MessageMediaPhoto, MessageMediaDocument
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configuration
API_ID = os.getenv('TELEGRAM_API_ID')
API_HASH = os.getenv('TELEGRAM_API_HASH')
PHONE = os.getenv('TELEGRAM_PHONE')
GROUP_ID = os.getenv('GROUP_ID') or os.getenv('MAIN_GROUP_ID')

# Session name
SESSION_NAME = 'cleanup_session'

# Statistics
stats = {
    'checked': 0,
    'deleted': 0,
    'kept_media': 0,
    'kept_today': 0,
    'errors': 0
}

def is_today(message_date):
    """Check if message is from today"""
    today = datetime.now().date()
    msg_date = message_date.date()
    return msg_date == today

def has_media(message):
    """Check if message contains media"""
    return (
        message.photo is not None or
        message.video is not None or
        message.document is not None or
        message.audio is not None or
        message.voice is not None or
        message.video_note is not None or
        message.sticker is not None or
        message.gif is not None
    )

async def cleanup_messages():
    """Main cleanup function"""
    print('\n🧹 Group Text Message Cleanup\n')

    # Check credentials
    if not API_ID or not API_HASH:
        print('❌ Error: Missing Telegram API credentials')
        print('\nPlease set the following environment variables:')
        print('   TELEGRAM_API_ID=your_api_id')
        print('   TELEGRAM_API_HASH=your_api_hash')
        print('\nGet them from: https://my.telegram.org\n')
        return

    if not GROUP_ID:
        print('❌ Error: Missing GROUP_ID in environment variables')
        return

    # Initialize client
    try:
        client = TelegramClient(SESSION_NAME, int(API_ID), API_HASH)
        await client.start(phone=PHONE or input('Enter your phone number: '))

        print(f'✅ Logged in successfully')
        print(f'📍 Target Group ID: {GROUP_ID}')
        print(f'📅 Today: {datetime.now().strftime("%Y-%m-%d")}\n')

        # Get the group
        try:
            entity = await client.get_entity(int(GROUP_ID))
            print(f'✅ Found group: {entity.title}\n')
        except Exception as e:
            print(f'❌ Error: Could not find group with ID {GROUP_ID}')
            print(f'   Error: {e}\n')
            return

        # Confirm action
        print('⚠️  WARNING: This will delete all text-only messages except from today!')
        print('   Media messages will be preserved.\n')
        confirm = input('Type "yes" to continue: ')

        if confirm.lower() != 'yes':
            print('❌ Cancelled by user\n')
            return

        print('\n🔍 Scanning messages...\n')

        # Fetch and process messages
        messages_to_delete = []
        batch_count = 0

        async for message in client.iter_messages(entity, limit=None):
            stats['checked'] += 1

            # Check if message has media
            if has_media(message):
                stats['kept_media'] += 1
                continue

            # Check if message is from today
            if is_today(message.date):
                stats['kept_today'] += 1
                continue

            # Check if message has text
            if message.text or message.message:
                messages_to_delete.append(message.id)
                date_str = message.date.strftime('%Y-%m-%d %H:%M')
                text_preview = (message.text or message.message)[:40]
                print(f'🗑️  Will delete: [{date_str}] {text_preview}...')

            # Delete in batches of 100
            if len(messages_to_delete) >= 100:
                try:
                    await client.delete_messages(entity, messages_to_delete)
                    stats['deleted'] += len(messages_to_delete)
                    batch_count += 1
                    print(f'\n✅ Deleted batch {batch_count} ({len(messages_to_delete)} messages)\n')
                    messages_to_delete = []
                    await asyncio.sleep(1)  # Rate limiting
                except Exception as e:
                    print(f'❌ Error deleting batch: {e}')
                    stats['errors'] += len(messages_to_delete)
                    messages_to_delete = []

        # Delete remaining messages
        if messages_to_delete:
            try:
                await client.delete_messages(entity, messages_to_delete)
                stats['deleted'] += len(messages_to_delete)
                print(f'\n✅ Deleted final batch ({len(messages_to_delete)} messages)\n')
            except Exception as e:
                print(f'❌ Error deleting final batch: {e}')
                stats['errors'] += len(messages_to_delete)

        # Print summary
        print('\n📊 Cleanup Summary:')
        print('─' * 50)
        print(f'   ✅ Total checked: {stats["checked"]}')
        print(f'   🗑️  Deleted: {stats["deleted"]}')
        print(f'   📸 Kept (media): {stats["kept_media"]}')
        print(f'   📅 Kept (today): {stats["kept_today"]}')
        print(f'   ❌ Errors: {stats["errors"]}')
        print('─' * 50)
        print('\n✅ Cleanup complete!\n')

    except Exception as e:
        print(f'\n❌ Error: {e}\n')
    finally:
        if 'client' in locals():
            await client.disconnect()

if __name__ == '__main__':
    asyncio.run(cleanup_messages())

File: scripts/cleanup-text-messages.js
#!/usr/bin/env node
/**
 * Group Text Message Cleanup Script
 *
 * This script deletes all text-only messages from a Telegram group,
 * except for messages sent today. Media messages are preserved.
 *
 * Usage: node scripts/cleanup-text-messages.js <group_id>
 * Example: node scripts/cleanup-text-messages.js -1001234567890
 */

require('dotenv-safe').config({ allowEmptyValues: true });
const { Telegraf } = require('telegraf');
const logger = require('../src/utils/logger');

// Configuration
const BOT_TOKEN = process.env.BOT_TOKEN;
const DEFAULT_GROUP_ID = process.env.GROUP_ID || process.env.MAIN_GROUP_ID;

// Get group ID from command line or use default
const GROUP_ID = process.argv[2] || DEFAULT_GROUP_ID;

if (!BOT_TOKEN) {
  console.error('❌ Error: BOT_TOKEN not found in environment variables');
  process.exit(1);
}

if (!GROUP_ID) {
  console.error('❌ Error: No group ID provided');
  console.error('Usage: node scripts/cleanup-text-messages.js <group_id>');
  console.error('Example: node scripts/cleanup-text-messages.js -1001234567890');
  process.exit(1);
}

/**
 * Check if a message is text-only (no media)
 */
function isTextOnlyMessage(message) {
  // Check if message has any media
  const hasMedia = !!(
    message.photo ||
    message.video ||
    message.animation ||
    message.document ||
    message.audio ||
    message.voice ||
    message.video_note ||
    message.sticker ||
    message.location ||
    message.venue ||
    message.contact ||
    message.poll
  );

  // Has text but no media = text-only message
  return (message.text || message.caption) && !hasMedia;
}

/**
 * Check if a message is from today
 */
function isFromToday(messageDate) {
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const msgDate = new Date(messageDate * 1000);
  msgDate.setHours(0, 0, 0, 0);

  return msgDate.getTime() === today.getTime();
}

/**
 * Get message info for logging
 */
function getMessageInfo(message) {
  const date = new Date(message.date * 1000);
  const from = message.from?.username
    ? `@${message.from.username}`
    : message.from?.first_name || 'Unknown';

  let content = '';
  if (message.text) {
    content = message.text.substring(0, 50);
    if (message.text.length > 50) content += '...';
  } else if (message.caption) {
    content = '[Caption] ' + message.caption.substring(0, 40);
  }

  return {
    id: message.message_id,
    date: date.toLocaleString(),
    from,
    content
  };
}

/**
 * Delay function for rate limiting
 */
function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Main cleanup function
 */
async function cleanupTextMessages() {
  console.log('\n🧹 Group Text Message Cleanup Script\n');
  console.log(`📍 Target Group: ${GROUP_ID}`);
  console.log(`📅 Today's date: ${new Date().toLocaleDateString()}\n`);

  const bot = new Telegraf(BOT_TOKEN);

  try {
    // Get chat info
    console.log('🔍 Fetching chat information...');
    const chat = await bot.telegram.getChat(GROUP_ID);
    console.log(`✅ Connected to: ${chat.title || chat.username || GROUP_ID}`);
    console.log(`   Type: ${chat.type}\n`);

    // Check bot permissions
    console.log('🔒 Checking bot permissions...');
    const botMember = await bot.telegram.getChatMember(GROUP_ID, bot.botInfo.id);
    const canDelete = botMember.can_delete_messages || botMember.status === 'creator';

    if (!canDelete) {
      console.error('❌ Error: Bot does not have permission to delete messages');
      console.error('   Please make the bot an admin with "Delete messages" permission');
      process.exit(1);
    }
    console.log('✅ Bot has delete permissions\n');

    console.log('⚠️  WARNING: This will delete all text-only messages except from today!');
    console.log('   Media messages will be preserved.');
    console.log('\n⏳ Scanning messages (this may take a while)...\n');

    let deletedCount = 0;
    let skippedCount = 0;
    let mediaCount = 0;
    let todayTextCount = 0;
    let errorCount = 0;
    let lastMessageId = null;
    let batchCount = 0;
    const maxBatches = 1000; // Safety limit

    // Note: Telegram doesn't provide a direct way to fetch message history via bot API
    // This script would need to be enhanced with actual message fetching logic
    // For now, I'll provide the deletion logic framework

    console.log('⚠️  NOTE: This script requires manual message ID collection.');
    console.log('   The Telegram Bot API does not provide message history retrieval.');
    console.log('\n   To delete messages, you need to:');
    console.log('   1. Use a user account (not bot) to fetch message history');
    console.log('   2. Use Telegram\'s getHistory method from a client');
    console.log('   3. Or provide message IDs manually\n');

    console.log('💡 Alternative: Create a bot command that deletes messages as they arrive');
    console.log('   or use Telegram Desktop\'s "Select messages" and bulk delete feature.\n');

    // If we had message IDs, the deletion would work like this:
    // Example deletion logic (would need actual message IDs):
    /*
    for (const messageId of messageIds) {
      try {
        // Get message details (if stored in database)
        const message = await getMessageFromDatabase(messageId);

        if (isTextOnlyMessage(message) && !isFromToday(message.date)) {
          await bot.telegram.deleteMessage(GROUP_ID, messageId);
          deletedCount++;
          console.log(`🗑️  Deleted: ${getMessageInfo(message).content}`);
        }

        // Rate limiting: delay between deletions
        await delay(100); // 100ms delay
      } catch (error) {
        errorCount++;
        logger.error(`Error deleting message ${messageId}:`, error.message);
      }
    }
    */

    console.log('\n📊 Summary:');
    console.log(`   ✅ Deleted: ${deletedCount} text messages`);
    console.log(`   📸 Preserved: ${mediaCount} media messages`);
    console.log(`   📅 Kept: ${todayTextCount} text messages from today`);
    console.log(`   ⏭️  Skipped: ${skippedCount} messages`);
    console.log(`   ❌ Errors: ${errorCount} messages\n`);

    console.log('✅ Cleanup complete!\n');

  } catch (error) {
    console.error('\n❌ Error during cleanup:', error.message);
    logger.error('Cleanup error:', error);
    process.exit(1);
  }
}

/**
 * Alternative: Create a bot-based cleanup using stored message IDs
 */
async function cleanupWithStoredMessages() {
  console.log('\n🧹 Cleanup using stored message data\n');

  const bot = new Telegraf(BOT_TOKEN);

  try {
    // This would query your database for message IDs
    // Example: SELECT message_id, date, has_media FROM messages WHERE chat_id = GROUP_ID

    const { query } = require('../src/config/postgres');

    console.log('📊 Querying database for message history...');

    // Get all messages from the group
    const result = await query(`
      SELECT
        message_id,
        date,
        text,
        media_type
      FROM messages
      WHERE chat_id = $1
      ORDER BY date DESC
    `, [GROUP_ID]);

    if (!result || result.rows.length === 0) {
      console.log('ℹ️  No messages found in database');
      return;
    }

    console.log(`✅ Found ${result.rows.length} messages in database\n`);

    const today = new Date();
    today.setHours(0, 0, 0, 0);

    let deletedCount = 0;
    let skippedCount = 0;

    for (const row of result.rows) {
      try {
        const msgDate = new Date(row.date);
        msgDate.setHours(0, 0, 0, 0);

        const isFromToday = msgDate.getTime() === today.getTime();
        const hasMedia = row.media_type && row.media_type !== 'text';

        // Delete if: text-only AND not from today
        if (!hasMedia && !isFromToday) {
          await bot.telegram.deleteMessage(GROUP_ID, row.message_id);
          deletedCount++;
          console.log(`🗑️  Deleted message ${row.message_id} from ${msgDate.toLocaleDateString()}`);

          // Rate limiting
          await delay(100);
        } else {
          skippedCount++;
          const reason = hasMedia ? 'has media' : 'from today';
          console.log(`⏭️  Skipped message ${row.message_id} (${reason})`);
        }
      } catch (error) {
        if (error.response?.error_code === 400) {
          console.log(`⚠️  Message ${row.message_id} already deleted or not found`);
        } else {
          console.error(`❌ Error deleting message ${row.message_id}:`, error.message);
        }
      }
    }

    console.log('\n📊 Summary:');
    console.log(`   ✅ Deleted: ${deletedCount} text messages`);
    console.log(`   ⏭️  Skipped: ${skippedCount} messages\n`);

  } catch (error) {
    console.error('\n❌ Error:', error.message);
    logger.error('Cleanup error:', error);
  }
}

// Run the appropriate cleanup method
if (process.env.USE_DATABASE === 'true') {
  cleanupWithStoredMessages()
    .then(() => process.exit(0))
    .catch(err => {
      console.error('Fatal error:', err);
      process.exit(1);
    });
} else {
  cleanupTextMessages()
    .then(() => process.exit(0))
    .catch(err => {
      console.error('Fatal error:', err);
      process.exit(1);
    });
}
